odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+        # - Odd and shifted -> - stride / 2
+        # - Even and not shifted -> + stride / 2
+        last_odd = shape[axes[-1]] % 2 == 1
+        last_shifted = shift_list[-1]
+        half_rstride = np.pi / (shape[axes[-1]] * stride[axes[-1]])
+
+        if last_odd:
+            if last_shifted:
+                rmax[axes[-1]] = -half_rstride
+            else:
+                rmax[axes[-1]] = 0
+        else:
+            if last_shifted:
+                rmax[axes[-1]] = 0
+            else:
+                rmax[axes[-1]] = half_rstride
+
+    return RegularGrid(rmin, rmax, rshape)
+
+
+def realspace_grid(recip_grid, x0, axes=None, halfcomplex=False,
+                   halfcx_parity='even'):
+    """Return the real space grid from the given reciprocal grid.
+
+    Given a reciprocal grid::
+
+        xi[j] = xi[0] + j * sigma,
+
+    with a multi-index ``j = (j[0], ..., j[d-1])`` in the range
+    ``0 <= j < M``, this function calculates the original grid::
+
+        x[k] = x[0] + k * s
+
+    by using a provided ``x[0]`` and calculating the stride ``s``.
+
+    If the reciprocal grid is interpreted as coming from a usual
+    complex-to-complex FFT, it is ``N == M``, and the stride is::
+
+        s = 2*pi / (sigma * N)
+
+    For a reciprocal grid from a real-to-complex (half-complex) FFT,
+    it is ``M[i] = floor(N[i]/2) + 1`` in the last transform axis ``i``.
+    To resolve the ambiguity regarding the parity of ``N[i]``, the
+    it must be specified if the output shape should be even or odd,
+    resulting in::
+
+        odd : N[i] = 2 * M[i] - 1
+        even: N[i] = 2 * M[i] - 2
+
+    The output stride is calculated with this ``N`` as above in this
+    case.
+
+    Parameters
+    ----------
+    recip_grid : `RegularGrid`
+        Sampling grid in reciprocal space.
+    x0 : `array-like`
+        Desired minimum point of the real space grid.
+    axes : int or `sequence` of ints, optional
+        Dimensions in which to calculate the real space grid. The sequence
+        must have the same length as ``shift`` if the latter is given
+        as a sequence. ``None`` means "all axes".
+    halfcomplex : bool, optional
+        If ``True``, interpret the given grid as the reciprocal as used
+        in a half-complex FFT (see above). Otherwise, the grid is
+        regarded as being used in a complex-to-complex transform.
+    halfcx_parity : {'even', 'odd'}
+        Use this parity for the shape of the returned grid in the
+        last axis of ``axes`` in the case ``halfcomplex=True``
+
+    Returns
+    -------
+    irecip : `RegularGrid`
+        The inverse reciprocal grid
+    """
+    if axes is None:
+        axes = list(range(recip_grid.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    rstride = recip_grid.stride
+    rshape = recip_grid.shape
+
+    # Calculate shape of the output grid by adjusting in axes[-1]
+    irshape = list(rshape)
```

Comment:
Yes, we are. For some odd reason, `list` has a `copy()` method only on Python 3. This works for both and creates a copy, which I prefer to importing the `copy` module.